src/Compilers/CSharp/Portable/Syntax/SyntaxExtensions.cs

```diff
@@ -398,84 +400,115 @@ internal static bool IsOutVarDeclaration(this DeclarationExpressionSyntax p)
         /// <param name="argument">The argument that is passed to the action whenever it is invoked</param>
         internal static void VisitRankSpecifiers<TArg>(this TypeSyntax type, Action<ArrayRankSpecifierSyntax, TArg> action, in TArg argument)
         {
-recurse:
-            switch (type.Kind())
+            // Use a manual stack here to avoid deeply nested recursion which can blow the real stack
+            var stack = new Stack<ArrayRankSpecifierOrTypeSyntax>();
```

Comment:
Consider using `ArrayBuilder`. There are `Push` and `Pop` methods on that class.

Reply:
I always forget that can act as a stack, thanks.

---
In reply to: [421213185](https://github.com/dotnet/roslyn/pull/44040#discussion_r421213185) [](ancestors = 421213185)